### PR TITLE
bv-to-bool and bool-to-bv preprocessing passes

### PR DESCRIPTION
**Pull request by yoni206**
Splitted bv-to-bool.{h,cpp} files from theory/bv into bv_to_bool.{h,cpp} and bool_to_bv.{h,cpp} in preprocessing/passes.
Also, made necessary changes for the new preprocessing schemes.

### CODE REVIEW

File: src/preprocessing/passes/bv_to_bool.cpp [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+#include "smt_util/node_visitor.h"
+
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
+
+using namespace std;
+using namespace CVC4::theory;
+
+using NodeMap = std::unordered_map<Node, Node, NodeHashFunction>;
```
Comment (4tXJ7f) [inline src/preprocessing/passes/bv_to_bool.cpp]: This doesn't seem to be used -> remove.

File: src/smt/smt_engine.cpp [hunk 2]
```diff
@@ -1228,6 +1229,8 @@ SmtEngine::~SmtEngine()
     d_definedFunctions->deleteSelf();
     d_fmfRecFunctionsDefined->deleteSelf();
 
+    d_private->unregisterPreprocessingPasses();
```
Comment (4tXJ7f) [inline src/smt/smt_engine.cpp]: Can we add a comment here that says that this is to make sure that we destroy the passes before we destroy things that they refer to?

File: src/smt/smt_engine.cpp [hunk 3]
```diff
@@ -71,6 +71,8 @@
 #include "preprocessing/passes/bv_gauss.h"
 #include "preprocessing/passes/int_to_bv.h"
 #include "preprocessing/passes/pseudo_boolean_processor.h"
+#include "preprocessing/passes/bool_to_bv.h"
+#include "preprocessing/passes/bv_to_bool.h"
```
Comment (HanielB) [inline src/smt/smt_engine.cpp]: this is not in alphabetical order w.r.t. the other includes

File: src/preprocessing/passes/bv_to_bool.h [hunk 4]
```diff
@@ -16,71 +16,64 @@
...
     Statistics();
     ~Statistics();
-  };
-  
-  NodeNodeMap d_liftCache;
-  NodeNodeMap d_boolCache;
-  Node d_one;
-  Node d_zero;
-
+  };
```
Comment (HanielB) [inline src/preprocessing/passes/bv_to_bool.h]: trailing whitespaces

File: src/smt/smt_engine.cpp [hunk 5]
```diff
@@ -1224,6 +1225,9 @@ SmtEngine::~SmtEngine()
 
     d_definedFunctions->deleteSelf();
     d_fmfRecFunctionsDefined->deleteSelf();
+
```
Comment (HanielB) [inline src/smt/smt_engine.cpp]: trailing whitespaces

**Pull request merged**